### PR TITLE
Feat: add material base

### PR DESCRIPTION
**Pull request by zhuxudong**
### Please check if the PR fulfills these requirements

- [x] The commit message follows our [guidelines](https://github.com/oasis-engine/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)
feature.
### What is the current behavior? (You can also link to an open issue here)
use `isTransparent`,`blendMode`,`renderFace` with dif mateials.
### What is the new behavior (if this is a feature change)?
Use `isTransparent`,`blendMode`,`renderFace` with the base material.
### Does this PR introduce a breaking change? (What changes might users need to make in their application due to this PR?)
Yes.
* `baseColorTexture` -> `baseTexture`.
* delete `alphaMode` , `doubleSided` from all materials.
* add `isTransparent`,`blendMode`,`renderFace` to all materials.
* `alphaCutoff` will take effect automatically.
### Other information:

### CODE REVIEW

File: packages/core/src/material/BaseMaterial.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  private static _blendMacro: ShaderMacro = Shader.getMacroByName("ALPHA_BLEND");
+
+  private _isTransparent: boolean = false;
+  private _alphaCutoff: number = 0;
+  private _renderFace: RenderFace = RenderFace.Front;
+
+  /**
+   * Is this material transparent.
+   */
+  get isTransparent() {
```
Comment (GuoLei1990) [inline packages/core/src/material/BaseMaterial.ts:18]: add return type

Comment (GuoLei1990) [reply packages/core/src/material/BaseMaterial.ts:20]: Supplement comments " if true will render use blendMode".

File: packages/core/src/material/BaseMaterial.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
+import { BlendFactor, BlendOperation, CullMode, Shader } from "../shader";
+import { ShaderMacro } from "../shader/ShaderMacro";
+import { RenderFace } from "./enums/RenderFace";
+import { RenderQueueType } from "./enums/RenderQueueType";
+import { BlendMode } from "./enums/BlendMode";
+import { Material } from "./Material";
+
+export class BaseMaterial extends Material {
```
Comment (GuoLei1990) [inline packages/core/src/material/BaseMaterial.ts:8]: add abstract

File: packages/core/src/material/BaseMaterial.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    if (value > 0) {
+      this.shaderData.enableMacro(BaseMaterial._alphaCutoffMacro);
+      this.renderQueueType = RenderQueueType.AlphaTest;
+    } else {
+      this.shaderData.disableMacro(BaseMaterial._alphaCutoffMacro);
+      this.renderQueueType = this._isTransparent ? RenderQueueType.Transparent : RenderQueueType.Opaque;
+    }
+
+    this.shaderData.setFloat("u_alphaCutoff", value);
+  }
```
Comment (GuoLei1990) [inline packages/core/src/material/BaseMaterial.ts:72]: Here has a bug
```typescript
set alphaCutoff(value: number) {
    if (value === this._alphaCutoff) return;
    this._alphaCutoff = value;

    if (value > 0) {
      this.shaderData.enableMacro(BaseMaterial._alphaCutoffMacro);
      this.renderQueueType = this._isTransparent ? RenderQueueType.Transparent : RenderQueueType.AlphaTest;
    } else {
      this.shaderData.disableMacro(BaseMaterial._alphaCutoffMacro);
      this.renderQueueType = this._isTransparent ? RenderQueueType.Transparent : RenderQueueType.Opaque;
    }

    this.shaderData.setFloat("u_alphaCutoff", value);
  }
```

File: packages/core/src/material/BaseMaterial.ts [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+      this.blendMode = this._blendMode;
+      depthState.writeEnabled = false;
+      this.renderQueueType = RenderQueueType.Transparent;
+    } else {
+      this.shaderData.disableMacro(BaseMaterial._blendMacro);
+      targetBlendState.enabled = false;
+      depthState.writeEnabled = true;
+      this.renderQueueType = RenderQueueType.Opaque;
+    }
+  }
```
Comment (GuoLei1990) [inline packages/core/src/material/BaseMaterial.ts:47]:
```typescript
set isTransparent(value: boolean) {
    if (value === this._isTransparent) return;
    this._isTransparent = value;

    const {
      depthState,
      blendState: { targetBlendState }
    } = this.renderState;

    if (value) {
      this.shaderData.enableMacro(BaseMaterial._blendMacro);
      targetBlendState.enabled = true;
      this.blendMode = this._blendMode;
      depthState.writeEnabled = false;
      this.renderQueueType = RenderQueueType.Transparent;
    } else {
      this.shaderData.disableMacro(BaseMaterial._blendMacro);
      targetBlendState.enabled = false;
      depthState.writeEnabled = true;
      this.renderQueueType = this._alphaCutoff ? RenderQueueType.AlphaTest : RenderQueueType.Opaque;
    }
  }
```

**Review by GuoLei1990 — APPROVED** [review]
+1

**Comment by zhuxudong** [comment]
#29

**Pull request merged**